### ISSUE TITLE
Saturating ops for `Weight`

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#2086]
During work on `Block::weight` and `Transaction::weight` it was realized that overflowing ops are at risk to malicious usage but during sane usage doing overflow checks is unergonomic. We ended up using saturating ops.

We could potentially implement saturating ops for the `Weight` type and document the reasoning.

#### References:

- https://github.com/rust-bitcoin/rust-bitcoin/pull/2076#issuecomment-1729976190
- https://github.com/rust-bitcoin/rust-bitcoin/pull/2073#issuecomment-1729971977

**Comment by apoelstra** [comment]
I'm glad that @Kixunil has arrived again as we look into this, because it has the feel of an unwise idea. But I am in favor of it because:

* In practice transaction weights are typically in the low hundreds or thousands. A standard tx cannot exceed 400k weight and a tx in the blockchain cannot exceed 4M weight. So normal users are never within 1000x of overflowing u32, let alone u64.
* But internally we do need to do arithmetic and in theory a user can try to consruct a transaction with a gazillion inputs, or multi-terabyte scripts, or whatever. So it's difficult to avoid the possibility overflow. (We should put a max size invariant on `Script` I think, but it's less obvious that we can/should put a max number of inputs or outputs on `Transaction`. Though I observe that Bitcoin itself starts to break down if you somehow have more than 2^32 outputs.)
* So...the three ways to deal with overflow are to return a Result, panic, or saturate.
   * Returning a Result for a case that most users will never come close to hitting sucks. And it infects basically every `Weight` related method in the library.
   * Panicking in such a case is a DoS vector, since it *is* possible to hit if you really try, even if it's impossible if you don't.
   * Saturating will result in silently wrong values in the case that you provide a wildly invalid transaction, but have a clean API and no DoSes.

**Comment by Kixunil** [comment]
I'm not sure what is meant by "ops". If the traits then I really don't like going against Rust convention of panicking as that would be surprising. However I'm totally for having `.saturating_*()` methods.

**Comment by junderw** [comment]
I agree saturating all around and making it prominent in the docs is fine.

**Comment by tcharding** [comment]
By ops I meant ops traits `Add`, `Mul` for example. Just having methods may be better and not implementing the ops at all so that usage is exlicit?

**Comment by tcharding** [comment]
On further investigation this issue can probably be expanded to be "use saturating arithmetic for all types where over flow may lead to users writing software that allows attackers to steal money and where MAX should never really be anywhere close to being reached for all sane usage", tentatively I would include:
- `Amount` (currently panics for overflow in release mode), (need to think about `SignedAmount`)
- `Weight`
- `FeeRate`

We probably need types `VBytes`, `Bytes`, `VSize`, and `Size` (obviously we cannot use the names `Bytes` and `Size`) to support the same saturating behaviour.

**Comment by yancyribbens** [comment]
I'm not against adding `.saturating_*()` although I don't think it should be a replacement for `.checked_*()`.  To me there are different use cases depending on the situation.

**Comment by apoelstra** [comment]
It is easy to actually overflow `Amount`. We should follow Rust conventions there. You can add 4000 copies of MAX_MONEY and you've done it.

`FeeRate` I assumed was a floating point value. How can it overflow? Is it actually some sort of rational?

It is only `Weight` where people expect to follow the "C rules" where they can just blindly add and multiply and never expect overflow to actually happen. But having said this, I am warming up to @Kixunil's feeling that we should just panic, even for `Weight`, since this is the Rust convention, and if it's suprising to people then they're going to get similarly surprised by all the primitive int types.

But yes, we should implement `SaturatingAdd` and `CheckedAdd` and so on, in addition to our normal addition functions. (Not `WrappingAdd` since I don't think that makes sense for any of these types.)

**Comment by yancyribbens** [comment]
I think it makes sense to panic in cases where an overflow might happen instead of return a result or saturate.  Since an overflow is something we do not expect to happen, it should be handled differently then an error that is something we do expect.

cite: expected.https://doc.rust-lang.org/book/ch09-03-to-panic-or-not-to-panic.html

Also, using panic helps simplify the code a lot since infallible iterators don't exist for many types of iterators.

**Comment by apoelstra** [comment]
I think that if "panic in debug mode, overflow in release" is good enough for stdlib it should be good enough for us.

**Comment by yancyribbens** [comment]
> I think that if "panic in debug mode, overflow in release" is good enough for stdlib it should be good enough for us.

Seems reasonable to me and would simplify things.  Maybe something in the README for downstream consumers to beware since the responsibility of bounds checking will be theirs.
